inference-engine/src/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp

```diff
@@ -0,0 +1,122 @@
+#include <transformations/utils/utils.hpp>
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::SharedShapeOf, "SharedShapeOf", 0);
+
+bool ngraph::pass::SharedShapeOf::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    RUN_ON_FUNCTION_SCOPE(SharedShapeOf);
+    bool graph_rewritten = false;
+
+    std::map<ngraph::Output<Node>, std::vector<std::shared_ptr<ngraph::opset1::ShapeOf>>> source_to_shape_of;
+    std::map<ngraph::Output<Node>, std::vector<std::shared_ptr<ngraph::opset3::ShapeOf>>> source_to_shape_of_3;
+    for (const auto & node : f->get_ordered_ops()) {
+        // Recursively apply transformation for sub-graph based operations
+        if (auto sub_graph_node = std::dynamic_pointer_cast<op::util::SubGraphOp>(node)) {
+            if (auto sub_graph = sub_graph_node->get_function()) {
+                graph_rewritten |= run_on_function(sub_graph);
+            }
+        }
+        if (auto shape_of = std::dynamic_pointer_cast<ngraph::opset1::ShapeOf>(node)) {
+            source_to_shape_of[shape_of->input_value(0)].push_back(shape_of);
+        } else if (auto shape_of_3 = std::dynamic_pointer_cast<ngraph::opset3::ShapeOf>(node)) {
+            source_to_shape_of_3[shape_of_3->input_value(0)].push_back(shape_of_3);
+        }
+    }
+
+    for (const auto& pair : source_to_shape_of) {
+        if (pair.second.size() < 2)
+            continue;
+        auto root_ss = pair.second[0];
+        for (auto& child_ss : pair.second) {
+            if (root_ss->get_instance_id() != child_ss->get_instance_id()) {
+                graph_rewritten |= replace_output_update_name(child_ss->output(0), root_ss->output(0));
+            }
+        }
+    }
+    for (const auto& pair : source_to_shape_of_3) {
+        if (pair.second.size() < 2)
+            continue;
+        auto root_ss = pair.second[0];
+        for (auto& child_ss : pair.second) {
+            if (root_ss->get_instance_id() != child_ss->get_instance_id() && root_ss->get_output_type() == child_ss->get_output_type()) {
+                graph_rewritten |= replace_output_update_name(child_ss->output(0), root_ss->output(0));
+            }
+        }
+    }
+    return graph_rewritten;
+}
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::GroupedGatherElimination, "GroupedGatherElimination", 0);
+
+bool ngraph::pass::GroupedGatherElimination::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    RUN_ON_FUNCTION_SCOPE(GroupedGatherElimination);
+    bool graph_rewritten = false;
+
+    for (const auto & node : f->get_ordered_ops()) {
+        // Recursively apply transformation for sub-graph based operations
+        if (auto sub_graph_node = std::dynamic_pointer_cast<op::util::SubGraphOp>(node)) {
+            if (auto sub_graph = sub_graph_node->get_function()) {
+                graph_rewritten |= run_on_function(sub_graph);
+            }
+        }
+        if (auto concat = std::dynamic_pointer_cast<ngraph::opset1::Concat>(node)) {
```

Comment:
Can we use matcher pass here?

Reply:
Like for the Concat operation only? isn't it an overkill? and why do we need that?

Reply:
Because you won't perform any manual graph traversals. It simplifies code. Also this pass can be included into GraphRewrite if needed. We use FunctionPass only in rare cases like in your case with shared ShapeOf optimization.